index.rst

```diff
@@ -447,6 +447,64 @@ in some contexts where materializing the set would be expensive. The
 'DataRefSet' is usually generated by listing existing datasets of a particular
 dataset type, but its component 'DataRef's can be used with other dataset types.
 
+Caching
+-------
+
+When an object is read by the Butler, the Butler keeps a weakref to that object
+and returns a normal reference to that object. As long as the normal ref is kept
+(i.e. as long as the object is not garbage collected) the Butler will keep the
+weakref. If that object is read again, the butler will get a normal ref (via the
+weakref) from the cache instead of reading the object from persisted data a
+second time.
+
+This is accomplished by keeping a WeakValueDictionary in Butler, where the key
+is a custom hash of the ButlerLocation passed to Butler._read, and the value is
+the object that was loaded using that ButlerLocation. The hash of ButlerLocation
+parameters includes the items needed to uniquely identify an object to be loaded
+(or retrieved from the cache).
```

Comment:
good point about datasetType being a complete key, pythonType and cppType are redundant. I think it's worth keeping a reference about which repo (in the form of the storage pointer) there; I'd like to be certain that a matching dataset from a different repo won't be loaded.